[stable27] Allow share expiry dates lower than enforced limits

nfebe opened the pull request:
Manual backport : https://github.com/nextcloud/server/pull/40927

nfebe committed:
Allow share expiry dates lower than enforced limits

Previously, users could change the share expiry date up
to the enforced maximum.

The new share flow imposed the enforced share expiry
date maximum literally and did not allow even dates
lower than the maximum enforced.

That does not make much sense, if the enforced expiry date is 30
days from creation date, then it's logical to allow users set the
date to anything less than 30 days from the creation date.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

come-nc commented on apps/files_sharing/src/views/SharingDetailsTab.vue:
```diff
@@ -418,11 +417,16 @@ export default {
-				return new Date(new Date().setDate(new Date().getDate() + 1 + this.config.defaultInternalExpireDate))
-			} else if (this.config.isDefaultRemoteExpireDateEnforced) {
-				return new Date(new Date().setDate(new Date().getDate() + 1 + this.config.defaultRemoteExpireDate))
+		maxExpirationDateEnforced() {
+			if (this.isPublicShare) {
+				return this.config.defaultExpirationDate
+			}
+			if (this.isRemoteShare) {
+				return this.config.defaultRemoteExpirationDateString
+			}
+			// If it get's here then it must be an internal share
+			if (this.isExpiryDateEnforced) {
+				return this.config.defaultInternalExpirationDate
```
Am I reading this wrong or is the date enforced even if the checkbox is not checked for public and remote shares?

Fenn-CS replied:
Nice catch @come-nc, please review : https://github.com/nextcloud/server/pull/40933 and I would back port it here before merge.

nfebe committed:
Check expiry enforcement for all share types

We should check that a maximum expiry date has been enforced for
all shares and NOT JUST FOR internal shares before enforcing a UI max date,
like in commit 9757e680e2f643fe19432255c4142e96fbc88df5

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

nfebe committed:
Fix has expiration date logic

Current expiration date errorneously assumes that `defaultExpirationDate`
applies to all kinds of shares. But it only really applies to public shares despite
its name.

This commit, fixes that by paring expiration dates with the correct share types during
new share initialization and simplifying the `hasExpirationDate` (check) property.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

nfebe committed:
Do not treat `isDefaultExpireDateEnforced` as fallback

`isDefaultExpireDateEnforced` and its corresponding `defaultExpirationDate`
is currently treated as the enforcement fallback when share type enforcements are not
set.

However, `isDefaultExpireDateEnforced` and `defaultExpirationDate` are actually more like
`isDefaultPublicExpireDateEnforced` and `defaultPublicExpirationDate` and therefore only applies
to public shares.

It might be ideal to rename this variables all the way from the backend config to the way we use
them in the frontend code.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>
Signed-off-by: nextcloud-command <nextcloud-command@users.noreply.github.com>

nfebe committed:
Fix dupblicate sharing option

Allow download was duplicated via a rebase/merge action during
fe78157b77675b2fdc9eb5abfe0d530727fdddd8 this fixes that.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

come-nc reviewed (approved):
Tested and works
